tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd

```diff
@@ -147,131 +164,79 @@ Note that:
 
 ```julia
-size(chain_ppca)
+size(chain_ppca) # (no. of iterations, no. of vars, no. of chains) = (500, 159, 1)
 ```
 
-The Chains struc `chain_ppca` also contains the sampling info such as rhat, ess, mean estimates, etc. You can print it to check these quantities.
-
-#### Step 5: diagnostics and posterior predictive checks
-
-Diagnostics of the inference results, i.e. convergence (r-hat), effective samples size (ess), acceptance rate, etc, can be performed using the `MCMCChain` package.
-Here we plot the sample trajectories for 1 chain:
+The Chains struc `chain_ppca` also contains the sampling info such as r-hat, ess, mean estimates, etc. You can print it to check these quantities.
 
-```julia
-plot(chain_ppca[:, 1, :]; xlabel="iteration_no", ylabel="sample_value")
-```
-
-We observe that, the samples are not clustered around a certain value, which indicates equilibrium may not achieve. This could be associated with the probabilistic model and the sampling settings used.
+#### Step 5: posterior predictive checks
 
 Posterior predictive check. We reconstruct the input data from our parameter estimates, using the posterior mean as parameter estimates.
-We first retrieve the samples for the projection matrix `w` from `chain_ppca`. This can be done using the Julia `group(chain, parameter name)` function.
-Then we calculate the mean value for each element in $w$, averaging over the whole chain of samples.
-
-> You can check that, the operation `group(chain_ppca, :w)` returns a chain with sahape $500 \times 18 \times 1$ (there are $2 \times 9$ elements in the projection matrix $w$), as expected.
-> The operation `mean(group(chain_ppca, :w))` gives an array with shape (18,2) - the first column is the parameter name, and the second column its mean estimate.
-> Then the vector of mean values is re-shaped into the original shape of $w$.
-> We repeat this for $z$ and $mu$ to calculate their mean estimates over the whole chain.
-> Finally, we obtain the predictions $x=w \times z + \mu_j$ where $\mu_j$ is specific to each original dimension $d \in D$.
-> The predictions for $x$, produced using the mean estimates of $w$, $z$ and $mu$ are then visualised.
+We first retrieve the samples for the projection matrix `W` from `chain_ppca`. This can be done using the Julia `group(chain, parameter_name)` function.
+Then we calculate the mean value for each element in $W$, averaging over the whole chain of samples.
 
 ```julia
 # Extract parameter estimates for predicting x - mean of posterior
-w = reshape(mean(group(chain_ppca, :w))[:, 2], (n_genes, k))
-z = permutedims(reshape(mean(group(chain_ppca, :z))[:, 2], (k, n_cells)))'
-mu = mean(group(chain_ppca, :m))[:, 2]
+W = reshape(mean(group(chain_ppca, :W))[:, 2], (n_genes, k))
+Z = reshape(mean(group(chain_ppca, :Z))[:, 2], (k, n_cells))
+μ = mean(group(chain_ppca, :μ))[:, 2]
 
-X = w * z .+ mu
+mat_rec = W * Z .+ repeat(μ; inner=(1, n_genes))'
```

Comment:
```suggestion
mat_rec = W * Z  .+ μ'
```

Reply:
Many thanks for the simplification. In the current setting, it becomes: ```mat_rec = W * Z .+ repeat(μ; inner=(1, n_cells))```
A bit hesitating between these two options:
1. mat_rec = W * Z .+ μ
2. mat_rec = W * Z .+ repeat(μ; inner=(1, n_cells))
The re-constructed `mat_rec`s are equivalent. The logic behind is that, if `μ` is a column vector, then `W * Z .+ μ` will add `μ` to each column (this is the case indicated by the results, and it is what we want); if `μ` is a row vector, then `W * Z .+ μ` will add `μ` to each row. So the addition Op `W * Z .+ μ` is executable due to this automatic broadcasting.
One may argue that, it is more readable and understandable to write matrix + matrix, rather than matrix + vector, that's why I prefer ```W * Z .+ repeat(μ; inner=(1, n_cells))```, although ```W * Z .+ μ``` looks simpler.